### PR TITLE
Fix duration measurement issues in the client

### PR DESCRIPTION
**Pull request by fmeum**
Even with `CLOCK_MONOTONIC`, we are still seeing Bazel servers fail to start up occasionally due to start time being larger than end time. Make this non-fatal by truncating to 0 and emitting a warning with start and end time to facilitate further investigation.

Also flip the conditions for command and extraction wait time, which previously were only included if *not* known.

### CODE REVIEW

**Comment by fmeum** [comment]
@bazel-io fork 8.0.0

**Review by tjgq — COMMENTED** [review]
High-level comment: I realize that I'm asking for a larger refactor, but I'm wondering if we could simply use `std::chrono::duration` for `DurationMillis`, and `std::optional<std::chrono::duration>` for `ExtractionDurationMillis` (because we either extracted and have a known duration, or we didn't extract and don't know it; the other two combinations don't appear to be useful).

WDYT?

File: src/main/cpp/blaze.cc [hunk 1]
```diff
@@ -557,14 +557,14 @@ static void AddLoggingArgs(const LoggingInfo &logging_info,
 
   // The time in ms a command had to wait on a busy Blaze server process.
   // This is part of startup_time.
-  if (command_wait_duration_ms.IsUnknown()) {
+  if (!command_wait_duration_ms.IsUnknown()) {
```
Comment (tjgq) [inline src/main/cpp/blaze.cc:560]: In the spirit of avoiding double negatives (which I suspect might have contributed to the bug) can we call the method `IsKnown()` instead?

File: src/main/cpp/archive_utils.h [hunk 2]
```diff
@@ -29,26 +30,12 @@ void DetermineArchiveContents(const std::string &archive_path,
...
-
-  bool IsUnknown() const { return millis == kUnknownDuration; }
-
- private:
-  // Value representing that a timing event never occurred or is unknown.
-  static constexpr uint64_t kUnknownDuration = 0;
-};
-
 // DurationMillis that tracks if an archive was extracted.
 struct ExtractionDurationMillis : DurationMillis {
```
Comment (tjgq) [inline src/main/cpp/archive_utils.h:34]: I wonder if we really need this class: we either did an extraction and know the time it took, or we didn't and we don't; we never inspect `archive_extracted`, except in tests. Wouldn't a `DurationMillis` suffice, with "unknown" signifying "not extracted"?

File: src/main/cpp/blaze_util_posix.cc [hunk 3]
```diff
@@ -731,7 +732,7 @@ LockHandle AcquireLock(const std::string& name, const blaze_util::Path& path,
     }
   }
 
-  *wait_time = elapsed_time;
+  *wait_time = elapsed_time->millis;
```
Comment (tjgq) [inline src/main/cpp/blaze_util_posix.cc:735]: Returning a `DurationMillis` through an out-parameter seems less awkward.

File: src/main/cpp/blaze.cc (hunk 1 above)
Comment (tjgq) [reply src/main/cpp/blaze.cc:560]: Thinking some more about it: if the only consequence of `IsUnknown()` is that we don't set a `--command_wait_time` flag, why not set it to 0 in that case? The flags library can't distinguish between 0 and unset anyway.

**Comment by fmeum** [comment]
@tjgq Just FYI, `blaze_archive_test.cc` doesn't seem to be included in any public test targets.

**Comment by tjgq** [comment]
> @tjgq Just FYI, `blaze_archive_test.cc` doesn't seem to be included in any public test targets.

Right. It looks like it needs a bit of non-trivial work to open source (it uses google3-only libraries) so I'm not going to do it for now.

We have an issue, though: in the google3 test environment, archive extraction often completes in under 1ms, which makes the tests extremely flaky. WDYT about adding `max(1, et-st)` to the `DurationMillis` constructor? We could also go back to having a separate sentinel for unknown, but I feel like it's more useful to make `--extract_data_time` a reliable signal on whether extraction took place.

**Comment by fmeum** [comment]
> On second thought - let's just go with std::optional<DurationMillis> (I can make the changes myself). 

Sounds good to me (and better than `max`)!

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 8.0.0 RC3](https://github.com/bazelbuild/bazel/releases/tag/8.0.0rc3). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=8.0.0rc3`. Thanks!

**Comment by fmeum** [comment]
@bazel-io fork 7.5.0

**Pull request closed without merging**